### PR TITLE
executor: add back metrics for select_result.go:fetchResp()

### PR DESCRIPTION
**Pull request by ichn-hu**
### What problem does this PR solve? 

In a previous PR https://github.com/pingcap/tidb/pull/13274 I accidentally removed a metric and some error handling code, here I added them back.

### What is changed and how it works?

Resume the `defer` function.

### Check List 

 - No code

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -74,14 +74,32 @@ type selectResult struct {
...
 	memTracker *memory.Tracker
 }
 
 func (r *selectResult) Fetch(ctx context.Context) {
 }
 
 func (r *selectResult) fetchResp(ctx context.Context) error {
+	startTime := time.Now()
+	defer func() {
+		if c := recover(); c != nil {
```
Comment (SunRunAway) [inline distsql/select_result.go]: We need not to do `recover` here since it is in main thread.

Comment (ichn-hu) [reply distsql/select_result.go]: You mean we could just remove the recovery part and let (if any) OOM panic goes up?

Comment (SunRunAway) [reply distsql/select_result.go]: Yes.

**Comment by ichn-hu** [comment]
@SunRunAway PTALA.

File: distsql/select_result.go [hunk 2]
```diff
@@ -81,6 +81,13 @@ func (r *selectResult) Fetch(ctx context.Context) {
 }
 
 func (r *selectResult) fetchResp(ctx context.Context) error {
+	startTime := time.Now()
+	defer func() {
+		duration := time.Since(startTime)
```
Comment (SunRunAway) [inline distsql/select_result.go]: Sorry for the roundtrip, but I think it's better to just gauge the time of `r.resp.Next` and exclude the deserialization time, same as previous PR #13274

Comment (ichn-hu) [reply distsql/select_result.go]: You are right, we should not just gauge the running time here, since it is not that crucial.

Comment (ichn-hu) [reply distsql/select_result.go]: I've updated the code, PTALA @SunRunAway

File: distsql/select_result.go [hunk 3]
```diff
@@ -74,28 +74,30 @@ type selectResult struct {
...
-	startTime := time.Now()
 	defer func() {
-		duration := time.Since(startTime)
-		r.fetchDuration += duration
-		if r.selectResp == nil {
+		if r.selectResp == nil && !r.durationReported {
 			// final round of fetch
 			// TODO: Add a label to distinguish between success or failure.
 			// https://github.com/pingcap/tidb/issues/11397
 			metrics.DistSQLQueryHistgram.WithLabelValues(r.label, r.sqlType).Observe(r.fetchDuration.Seconds())
```
Comment (SunRunAway) [inline distsql/select_result.go]: Could you put these into line 104, since it is not quite necessary using `defer` here.

Comment (SunRunAway) [reply distsql/select_result.go]: NVM, it's ok with me.

**Comment by ichn-hu** [comment]
@wshwsh12 PTAL

File: distsql/select_result.go [hunk 4]
```diff
@@ -92,6 +97,13 @@ func (r *selectResult) fetchResp(ctx context.Context) error {
 		}
 		if resultSubset == nil {
 			r.selectResp = nil
+			if !r.durationReported {
```
Comment (wshwsh12) [inline distsql/select_result.go]: Why need this check? Can't Observe the time immediately？

Comment (ichn-hu) [reply distsql/select_result.go]: Nope, the original metric is used for tracking the total running time of data fetching, therefore we have to keep the behavior the same with the original metric.

**Pull request merged**